framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,901 @@
+//  elements, but this currently is not recommended since the
+//  insert-annotations-to-source tool, which adds annotations from .jaif files
+//  into source code, adds annotations on top of existing
+//  annotations. See https://github.com/typetools/annotation-tools/issues/105 .
+//  TODO: Ensure that annotations are inserted deterministically into .jaif
+//  files. This is important otherwise developers might achieve different
+//  results (order of annotations) when running the signature inference for
+//  the same set of files.
+public class SignatureInferenceScenes {
+
+    /**
+     * Directory where .jaif files will be written to and read from.
+     * This directory is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build" + File.separator +
+            "signature-inference" + File.separator;
+
+    /** Maps .jaif file paths (Strings) to Scenes. */
+    private static Map<String, AScene> scenes = new HashMap<>();
+
+    /** Maps a TypeUseLocation to the name of all annotations that should not be
+     * added to .jaif files for that location.
+     */
+    private static Map<TypeUseLocation, Set<String>> annosToIgnore = new HashMap<>();
+
+    /**
+     * Set representing Scenes that were modified since the last time all
+     * Scenes were written into .jaif files. Each String element of this set
+     * is a path to the .jaif file of the corresponding Scene in the set. It
+     * is obtained by passing a class name as argument to the
+     * {@link #getJaifPath} method.
+     * <p>
+     * Modifying a Scene means adding (or changing) a type annotation for a
+     * field, method return type, or method parameter type in the Scene.
+     * (Scenes are modified by the method {@link #updateAnnotationSetInScene}.)
+     */
+    private static Set<String> modifiedScenes = new HashSet<>();
+
+    /**
+     * Returns the String representing the .jaif path of a class given its name.
+     */
+    private static String getJaifPath(String className) {
+        String jaifPath = JAIF_FILES_PATH + className + ".jaif";
+        return jaifPath;
+    }
+
+    /**
+     * Returns the Scene stored in a .jaif file path passed as input.
+     * If the file does not exist, an empty Scene is created.
+     */
+    private static AScene getScene(String jaifPath) {
+        if (!scenes.containsKey(jaifPath)) {
+            File jaifFile = new File(jaifPath);
+            AScene scene = new AScene();
+            if (jaifFile.exists()) {
+                try {
+                    IndexFileParser.parseFile(jaifPath, scene);
+                } catch (IOException e) {
+                    ErrorReporter.errorAbort("Problem while reading file in: " + jaifPath + "."
+                            + " Exception message: " + e.getMessage());
+                }
+            }
+            scenes.put(jaifPath, scene);
+        }
+        return scenes.get(jaifPath);
```

Comment:
The `scenes.get is redundant if you just added it. I would move the `AScene scene` before the if and initialize it in both branches, then return it.